cli/mender-artifact/dump.go

```diff
@@ -0,0 +1,242 @@
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/mendersoftware/mender-artifact/areader"
+	"github.com/mendersoftware/mender-artifact/artifact"
+	"github.com/mendersoftware/mender-artifact/handlers"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+type artifactDump struct {
+	// Reconstructed arguments from headers and content
+	args []string
+}
+
+type dumpFileStore struct {
+	fileDir string
+	args    *[]string
+}
+
+func DumpCommand(c *cli.Context) error {
+	var dump artifactDump
+
+	if len(c.Args()) != 1 {
+		return cli.NewExitError("Need to specify exactly one Artifact with dump command",
+			errArtifactInvalidParameters)
+	}
+
+	art, err := os.Open(c.Args()[0])
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf(
+			"Error opening Artifact: %s", err.Error()),
+			errArtifactOpen)
+	}
+	defer art.Close()
+
+	ar := areader.NewReader(art)
+
+	scriptsReadCallback := func(r io.Reader, i os.FileInfo) error {
+		fullPath := path.Join(c.String("scripts"), i.Name())
+		script, err := os.OpenFile(fullPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
+		if err != nil {
+			return err
+		}
+		defer script.Close()
+
+		_, err = io.Copy(script, r)
+		if err != nil {
+			return err
+		}
+
+		dump.args = append(dump.args, "--script", fullPath)
+
+		return nil
+	}
+	if len(c.String("scripts")) > 0 {
+		err = os.MkdirAll(c.String("scripts"), 0755)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf(
+				"Could not create directory: %s", err.Error()), errSystemError)
+		}
+		ar.ScriptsReadCallback = scriptsReadCallback
+	}
+
+	err = ar.ReadArtifactHeaders()
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Error dumping Artifact: %s",
+			err.Error()), errArtifactInvalid)
+	}
+
+	err = dumpPayloads(c, ar, &dump)
+	if err != nil {
+		return err
+	}
+
+	if c.Bool("print-cmdline") {
+		printCmdline(ar, dump.args)
+	}
+
+	return nil
+}
+
+func dumpPayloads(c *cli.Context, ar *areader.Reader, dump *artifactDump) error {
+	handlers := ar.GetHandlers()
+	if len(handlers) != 1 {
+		return cli.NewExitError("Only exactly one payload is supported for dump command",
+			errArtifactUnsupportedFeature)
+	}
+
+	if len(c.String("meta-data")) > 0 {
+		err := os.MkdirAll(c.String("meta-data"), 0755)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf(
+				"Unable to create directory: %s", err.Error()), errSystemError)
+		}
+
+		// Hardcode to 0 index for now.
+		handler := handlers[0]
+
+		for _, augmented := range []bool{false, true} {
```

Comment:
Clever, trick! But wouldn't a method be nice? And save someone else from wrangling with gocyclo at a later time?

Reply:
Yes, you're right. I'll move the meta-data stuff to their own function.